src/editor/components/modals/file-modal.jsx

```diff
@@ -0,0 +1,472 @@
+            file: unsavedFile
+          };
+          return newState;
+        });
+
+        // Attempt to upload the file and update state accordingly
+        const fileReader = new FileReader();
+        fileReader.readAsArrayBuffer(unsavedFile);
+        fileReader.onload = () => {
+          saveFileToServer(
+            this.props.notebookID,
+            fileReader.result,
+            unsavedFile.name
+          )
+            .then(savedFile => {
+              this.setState(state => {
+                const newState = state;
+                newState.files[key] = {
+                  status: "saved",
+                  file: savedFile
+                };
+                return newState;
+              });
+            })
+            .catch(() => {
+              this.setState(state => {
+                const newState = state;
+                newState.files[key].status = "error";
+                newState.files[key].errorMessage = "Upload error";
+                return newState;
+              });
+            });
+        };
+      }
+    });
+  };
+
+  /**
+   * Click the hidden file <input>, which will cause the browser's file
+   * selection dialog to appear.
+   */
+  openSelectDialog = () => {
+    this.fileInputRef.current.click();
+  };
+
+  /**
+   * Copy the path of the specified file to the clipboard and update state to
+   * reflect that this has been done recently so that a notice can be shown.
+   */
+  copyFilePath = id => {
+    // Copy the path to the clipboard
+    const pathElement = document.querySelector(`#file-${id} .file-path`);
+    pathElement.select();
+    document.execCommand("copy");
+
+    // If we don't do this and the user clicks many different "copy path"
+    // buttons in a row, the last notice to appear will disappear very quickly.
+    if (this.copiedFeedbackTimeout) {
+      clearTimeout(this.copiedFeedbackTimeout);
+    }
+
+    // Show the notice...
+    this.setState(state => {
+      const newState = state;
+      newState.pathJustCopied = id;
+      return newState;
+    });
+
+    // ...but hide it after some period of time.
+    this.copiedFeedbackTimeout = setTimeout(() => {
+      this.setState(state => {
+        const newState = state;
+        newState.pathJustCopied = null;
+        return newState;
+      });
+    }, 3000);
+  };
+
+  showDeleteConfirmationModal = (key, id, name) => {
+    this.setState({ pendingDelete: { key, id, name } });
+  };
+
+  hideDeleteConfirmationModal = () => {
+    this.setState({ pendingDelete: false });
+  };
+
+  /**
+   * Delete the file that the user confirmed they wanted to delete.
+   */
+  executePendingDelete = () => {
+    // setState enqueues state changes; they might not take effect immediately.
+    // To ensure that the delete modal disappears at the same time that the file
+    // disappears from the list, it's important to update "files" and
+    // "pendingDelete" in the same batch of setState updates.
```

Comment:
> can't both of these changes be enqueued with two calls to `setState` within `executePendingDelete`? They both should get enqueued directly after each other and then update accordingly.

I'm trying to think... if a user begins uploading a file, then deletes a file, which would cause this method to execute, isn't it possible that the file will finish uploading (causing a `setState` call) between the two `setState` calls in this method?

Reply:
Or are callbacks never executed in the middle of the execution of another method? I guess I've never thought about that.